### PR TITLE
cmd/ncdu: fix screen corruption

### LINKED ISSUE #6271: rclone ncdu: Fix screen corruption when logs happen
#### The associated forum post URL from `https://forum.rclone.org`

This was discussed in https://github.com/rclone/rclone/pull/6259 which was an attempt to fix this problem.

#### What problem are you are trying to solve?

Currently the `rclone ncdu` screen gets corrupted if rclone logs anything. This can be fixed with CTRL-L but it would be nice if it didn't happen at all.

#### How do you think rclone should be changed to solve that?

After a bit of discussion the best approach is probably

- capture the logging in the same way `--progress` does
    - https://github.com/rclone/rclone/blob/7a909ebfb0d100187fab7949455bf74b8ce07042/cmd/progress.go#L29-L40
- store the logs as they come in up to a maximum of 1000 (say)
- show the last log in the status bar
- show the logs in the buffer when exiting ncdu

At some point we could also make a log viewing window which would allow you to view / scroll the logs so far.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

### PR DESCRIPTION
**Pull request by eNV25**
#### What is the purpose of this change?

Fixes #6271

This PR implements two different approaches to fixing the issue, mentioned in the issue.

#### Was the change discussed in an issue or in the forum before?

Issue #6271

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by eNV25** [comment]
TIP: To create a warning on purpose, you can use a Google Drive backend with a dangling shortcut.

**Review by ncw — APPROVED** [review]
Sorry this has taken me a while to review.

I think this code looks good and I think it will work correctly on non terminfo systems, so I will merge it now for 1.63 - thank you :-)

**Comment by ncw** [comment]
Actually, maybe I won't merge it since you marked it as EXPERIMENTAL in the commit message :-)

I'm happy with the approach - the alternate screen is neat on terminfo systems and the printing the log messages at the end is OK on the others.

Anything else you want to do to the commit before merge?

**Comment by eNV25** [comment]
I don't know if you have noticed already, but the terminfo approach causes some screen flickering. I personally don't think it's too bad, but if you have many logs it pretty noticeable.

To mitigate it, I tried adding [synchronized update escape codes](https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036), but it didn't seem to help in kitty. I did NOT add this in the latest change, but if you want that anyway I can add it.

If you think this is not an issue, you can merge as-is.

The last change if only a small cleanup.

**Comment by eNV25** [comment]
> I gave it a bit more testing.
> 
> I noticed that it deadlocks after I press `q` when running
> 
> ```
> rclone ncdu -vv . --exclude "*"
> ```
> 

I see it too. I don't why that is, but I'll check it out with dlv.

> Using ubuntu 22.04 and xfce4-terminal
> 
> Which makes me think the double buffer term trick isn't ready for the big time.
> 
> I guess the alternative would be to do all platforms the same way by storing the log then printing it at the end so maybe we should merge that bit of code now and leave the double buffer trick?
> 
> This maybe has the advantage that we could display the last log on the screen
> 

I agree, the array option seems like a safer choice.

> > To mitigate it, I tried adding [synchronized update escape codes](https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036), but it didn't seem to help in kitty. I did NOT add this in the latest change, but if you want that anyway I can add it.
> 
> How many terminals support those codes?

Not many terminal support this. And there are two competing sets of escape code. But tmux has a terminfo extension for that, see the `Sync` option in `man tmux`. tcell's terminfo library doesn't support that, so it requires upstream changes.

File: cmd/ncdu/ncdu.go [hunk 1]
```diff
@@ -911,6 +912,27 @@ func (u *UI) Show() error {
...
+		}
+		var logs []log
+		fs.LogPrint = func(level fs.LogLevel, text string) {
+			if len(logs) > 100 {
+				logs = logs[len(logs)-100:]
+			}
+			logs = append(logs, log{level: level, text: text})
+		}
+		defer func() {
+			for i := range logs {
```
Comment (ncw) [inline cmd/ncdu/ncdu.go:930]: This function should also replace `fs.LogPrint = logPrint` I think?

**Review by ncw — COMMENTED** [review]
Looks good - thank you :-)

See query inline.

Let's get this merged first then we can do a followup with fancy term stuff if we think we can get it to work reliably.

File: cmd/ncdu/ncdu.go (hunk 1 above)
Comment (eNV25) [reply cmd/ncdu/ncdu.go:931]: True. I thought I pushed that, but it turns out I missed it.

**Review by ncw — APPROVED** [review]
Perfect - thank you :-)

**Pull request merged**